File: it-agents/skills/code-architect/references/code-examples/swift/Bridge.swift
// Bridge — split abstraction and implementation into independent hierarchies.

// Implementation hierarchy.
protocol Renderer {
    func drawShape(_ name: String) -> String
}

struct VectorRenderer: Renderer {
    func drawShape(_ name: String) -> String { "drawing \(name) as crisp vectors" }
}

struct RasterRenderer: Renderer {
    func drawShape(_ name: String) -> String { "drawing \(name) as pixels" }
}

// Abstraction hierarchy holds a reference to an implementation.
class Shape {
    let renderer: Renderer
    init(renderer: Renderer) { self.renderer = renderer }
    func draw() -> String { fatalError("subclass must override draw()") }
}

final class Circle: Shape {
    override func draw() -> String { renderer.drawShape("circle") }
}

final class Square: Shape {
    override func draw() -> String { renderer.drawShape("square") }
}

func runDemo() {
    let shapes: [Shape] = [
        Circle(renderer: VectorRenderer()),
        Circle(renderer: RasterRenderer()),
        Square(renderer: VectorRenderer())
    ]
    shapes.forEach { print($0.draw()) }
}

runDemo()

File: it-agents/skills/code-architect/references/code-examples/python/bridge.py
"""Bridge — split abstraction and implementation into independent hierarchies."""

from abc import ABC, abstractmethod


class Renderer(ABC):
    """Implementation side: how things are drawn."""

    @abstractmethod
    def draw_line(self, label: str) -> str:
        ...


class AsciiRenderer(Renderer):
    def draw_line(self, label: str) -> str:
        return f"--- {label} ---"


class HtmlRenderer(Renderer):
    def draw_line(self, label: str) -> str:
        return f"<hr/><b>{label}</b>"


class View(ABC):
    """Abstraction side: holds a reference to a Renderer."""

    def __init__(self, renderer: Renderer) -> None:
        self._renderer = renderer

    @abstractmethod
    def show(self) -> str:
        ...


class TitleView(View):
    def __init__(self, renderer: Renderer, title: str) -> None:
        super().__init__(renderer)
        self._title = title

    def show(self) -> str:
        return self._renderer.draw_line(self._title)


if __name__ == "__main__":
    print(TitleView(AsciiRenderer(), "Welcome").show())
    print(TitleView(HtmlRenderer(), "Welcome").show())
